### PR TITLE
[fix] Fixed subnet /32 & /128 pie chart error #129

### LINKED ISSUE #129: [bug] Subnet /32 gives error in pie chart
Try creating a /32 subnet.

Expected result:

- pie chart should show normally

Actual result:

Screenshot from 2022-01-05 11-38-52

### PR DESCRIPTION
**Pull request by Aryamanz29**
- Fixed `subnet /32` pie chart.
Fixes #129

See comment https://github.com/openwisp/openwisp-ipam/issues/129#issuecomment-1048060911

Screenshot from 2022-02-20 23-47-07

### CODE REVIEW

**Comment by Aryamanz29** [comment]
Coverage is decreasing, I need to work on this 😅

**Comment by Aryamanz29** [comment]
> Coverage is decreasing, I need to work on this sweat_smile

Done 👍

File: openwisp_ipam/api/views.py [hunk 1]
```diff
@@ -137,10 +137,10 @@ def count(self):
         if self.stop is not None:
             return self.stop - self.start
         broadcast = int(self.subnet.subnet.broadcast_address)
-        # IPV4 (exclude broadcast)
         if self.subnet.subnet.max_prefixlen == 32:
```
Comment (pandafy) [inline openwisp_ipam/api/views.py:140]:
```suggestion
        if self.subnet.subnet.version == 4:
```

I was wrong with my initial findings. The code is using `max_prefixlen` to differentiate between IPv4 and IPv6 networks. Let's use the `version` attribute for this instead.

File: openwisp_ipam/api/views.py [hunk 2]
```diff
@@ -137,10 +137,10 @@ def count(self):
...
+            if self.subnet.subnet._prefixlen == 32:
```
Comment (pandafy) [inline openwisp_ipam/api/views.py:141]:
```suggestion
            if self.subnet.subnet.prefixlen == 32:
```

`_prefixlen` is a private attribute. We should avoid using it. Use `prefixlen` instead. 
Also, add a comment here explaining why `/32` is handled separately.

File: openwisp_ipam/api/views.py [hunk 3]
```diff
@@ -137,10 +137,10 @@ def count(self):
...
+                return 1
             return broadcast - self.network - 1
-        # IPV6
```
Comment (pandafy) [inline openwisp_ipam/api/views.py:143]: Let's keep the initial code comments. Or you can improve them.

**Review by pandafy — CHANGES_REQUESTED** [review]
@Aryamanz29 there's no new test for testing this fix explicitly. Can you add a test that fails without this fix?

**Comment by Aryamanz29** [comment]
@pandafy Done requested changes 👍

File: openwisp_ipam/tests/test_admin.py [hunk 4]
```diff
@@ -426,3 +426,12 @@ def assert_response(response):
...
             assert_response(response)
+
+    def test_subnet_32(self):
+        subnet = self._create_subnet(subnet='192.168.0.0/32', description='Subnet 32')
+        try:
+            response = self.client.get(reverse('ipam:hosts', args=(subnet.id,)))
+            self.assertIsNone(response.data['next'])
+            self.assertIsNone(response.data['previous'])
+        except ValueError as err:
+            self.fail(f'subnet/32: {err}')
```
Comment (devkapilbansal) [inline openwisp_ipam/tests/test_admin.py:437]: I don't think we need a try-except block here :eyes: Tests will fail in case of error

Comment (Aryamanz29) [reply openwisp_ipam/tests/test_admin.py:437]: Right

**Review by nemesisdesign — COMMENTED** [review]
@Aryamanz29 the solution work and the code is good. I haven't tried the test yet but I will.

I have one request: I am not sure why I didn't test the same for IPv6 too but it seems to me that the problem happens there too with /128 subnets, can you fix that as well please? The solution and test used can be equivalent.

**Comment by Aryamanz29** [comment]
> @Aryamanz29 the solution work and the code is good. I haven't tried the test yet but I will.
> 
> I have one request: I am not sure why I didn't test the same for IPv6 too but it seems to me that the problem happens there too with /128 subnets, can you fix that as well please? The solution and test used can be equivalent.

Yes, I'll check this /128 subnets and update here :+1:

**Comment by Aryamanz29** [comment]
> > @Aryamanz29 the solution work and the code is good. I haven't tried the test yet but I will.
> > I have one request: I am not sure why I didn't test the same for IPv6 too but it seems to me that the problem happens there too with /128 subnets, can you fix that as well please? The solution and test used can be equivalent.
> 
> Sure, I'll check with /128 subnets and update +1

**Update :**
 Yes it's not showing pie chart with `subnet/128`, Also instead of raising `ValueError` at this endpoint `http://127.0.0.1:8000/api/v1/ipam/subnet/64d5dfc7-7845-42c1-8856-d3436dddc750/hosts/`, It is returning empty host list.

Screenshot from 2022-02-26 21-34-15,

**Comment by Aryamanz29** [comment]
@nemesisdesign 
This problem occurs due to this line :point_down: 
https://github.com/openwisp/openwisp-ipam/blob/4f15237ebe920987a1f81a3a7be6b7e0733c75a3/openwisp_ipam/api/views.py#L132
because of extra `+1` in host (But don't know why we're adding +1 here :thinking: ), It always shows the next address. After removing, It looks good :point_down: 

Screenshot from 2022-03-01 21-52-17

File: openwisp_ipam/api/views.py [hunk 5]
```diff
@@ -129,7 +129,12 @@ def __getitem__(self, i):
             return HostsSet(self.subnet, self.start + start, self.start + stop)
         if i >= self.count():
             raise IndexError
-        host = self.subnet.subnet._address_class(self.network + 1 + i + self.start)
+        # In case of single hosts ie subnet/32 & /128
+        if self.subnet.subnet.prefixlen in [32, 128]:
+            host = self.subnet.subnet._address_class(self.network + i + self.start)
+        else:
+            # Host starts from next address
+            host = self.subnet.subnet._address_class(self.network + 1 + i + self.start)
```
Comment (nemesisdesign) [inline openwisp_ipam/api/views.py:137]: great but we need a test for this new case please.

There's also some repetition here which can be avoided, avoid repeating the whole line `host = self.subnet.subnet._address_class(...)`

**Review by nemesisdesign — CHANGES_REQUESTED** [review]

File: openwisp_ipam/api/views.py (hunk 5 above)
Comment (Aryamanz29) [reply openwisp_ipam/api/views.py]: > great but we need a test for this new case please.
> 
> There's also some repetition here which can be avoided, avoid repeating the whole line `host = self.subnet.subnet._address_class(...)`

Done :+1:

File: openwisp_ipam/tests/test_admin.py [hunk 6]
```diff
@@ -426,3 +426,31 @@ def assert_response(response):
                 reverse('admin:ipam_export_subnet', args=[subnet.id]), follow=True
             )
             assert_response(response)
+
+    def test_subnet_32(self):
```
Comment (nemesisdesign) [inline openwisp_ipam/tests/test_admin.py:430]: Why are you adding these tests in `test_admin.py`, when the code for these tests is in `api.py`? These tests should be in `test_api.py`, please pay close attention to these details, we need to maintain the codebase clean and readable, I almost did not notice this.

**Review by nemesisdesign — CHANGES_REQUESTED** [review]
Good work @Aryamanz29, there's one outstanding issue which I highlighted below.

File: openwisp_ipam/tests/test_admin.py (hunk 6 above)
Comment (Aryamanz29) [reply openwisp_ipam/tests/test_admin.py]: My bad :worried: , moved tests to `test_api.py` :+1:

**Review by nemesisdesign — APPROVED** [review]
Thanks :+1:

**Pull request merged**